xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -184,8 +184,14 @@ final class WrrSubchannel extends ForwardingSubchannel {
 
     @VisibleForTesting
     void onLoadReport(MetricReport report) {
-      double newWeight = report.getCpuUtilization() == 0 ? 0 :
-              report.getQps() / report.getCpuUtilization();
+      double newWeight = 0;
+      if (report.getCpuUtilization() > 0 && report.getQps() > 0) {
+        double penalty = 0;
+        if (report.getEps() > 0 && config.errorUtilizationPenalty > 0) {
```

Comment:
The config object is not thread safety. It is not volatile anymore, but actually the config update and this onLoadReport referring to it might be from different threads.

Reply:
From my understanding, the entire `config` itself is only updated in `acceptResolvedAddresses()` and the fields (e.g. `config.errorUtilizationPenalty`) are final and cannot be modified/updated/set for all onLoadReport calls until another `acceptResolvedAddresses()` call.

Furthermore, `acceptResolvedAddresses()` then calls `afterAcceptAddresses()` which "re-register" the `onLoadReport` listeners. So the config field should not change for the lifetime of all `onLoadReport` calls?

Reply:
The config update is in `syncContext`. The perPRC `onLoadReport` listener  is set to stream tracer when pick subchannel, also in `syncContext`.  The config change seen from the other thread (the transport thread) has thread safety issue - this can be stale data or random data.

Note that the `afterAcceptAddress()` is only for oob listener, which is different from perPpcListener. OOB listener is fine, because it is called back from a `syncContext`.

cc. @ejona86

Reply:
Pick subchannel is not in the `syncContext`, and the tracer isn't called within the `syncContext`. The problem is not the contents of config, but reading the `config` field itself.

Reply:
That makes sense. I see that `config` is not `volatile` as of #10139, what would be the reasoning for not making it `volatile`?

Could a fix be (I think approach 1 would be better):

1.  add a separate `volatile` variable/field (in this class) for error utilization penalty and update this variable whenever the `config` is updated (after line 100-101). Although this does bring a slight cost of maintaining a `volatile` variable. 

2. use `synchronized` on the method but this incurs a heavy performance penalty with locking and since we are only reading the fields and `onLoadReport` could be called frequently, so **I don't think this would be a good approach**.

On a related note, I see that there are other reads of the `config` field in this file (e.g. in the `getWeight()` method), wouldn't those be currently affected as well since the `config` object itself is not `volatile`?

Reply:
Yes I'll correct that pick subchannel is not always in `syncContext` (it looks pending stream reprocess picking subchannel is in `syncContext`, but not directly pick when there is already a picker)

I agree that adding to the `errorUtilizationPenalty ` config to the  constructor is feasible. We already did that with other config (`config.enableOobLoadReport`)

Reply:
Other places like `getWeight()` is called within `syncContext` - the timer is scheduled from there so it is fine.

`per-rpc-listener` for the load reports is final per each subchannel, the stream tracer is constructed with the listener, so it looks fine for its thread safety.

Reply:
Done.
* Added `config.errorUtilizationPenalty` to the picker, then passed to the sub-channel, allocate a new listener each RPC (only for `perRpcListener` since OOB listener is already thread safe)
* Updated tests & improved coverage for edge cases not tested previously

Reply:
Please see updated commits & comment thread below.